File: 프로그래머스/2/160585. 혼자서 하는 틱택토/혼자서 하는 틱택토.js
function solution(board) {
    let countO = 0, countX = 0;

    // O, X 개수 세기
    for (let row of board) {
        for (let ch of row) {
            if (ch === 'O') countO++;
            else if (ch === 'X') countX++;
        }
    }

    // 승리 여부 판단 함수
    function checkWin(player) {
        for (let i = 0; i < 3; i++) {
            // 가로, 세로 체크
            if (board[i][0] === player && board[i][1] === player && board[i][2] === player) return true;
            if (board[0][i] === player && board[1][i] === player && board[2][i] === player) return true;
        }
        // 대각선 체크
        if (board[0][0] === player && board[1][1] === player && board[2][2] === player) return true;
        if (board[0][2] === player && board[1][1] === player && board[2][0] === player) return true;

        return false;
    }

    const oWin = checkWin('O');
    const xWin = checkWin('X');

    // 규칙 위반 조건들
    if (countX > countO) return 0;            // X가 더 많을 수 없음
    if (countO > countX + 1) return 0;        // O가 2개 이상 많을 수 없음
    if (oWin && xWin) return 0;               // 둘 다 승리 불가
    if (oWin && countO !== countX + 1) return 0; // O 승리인데 개수 안 맞음
    if (xWin && countO !== countX) return 0;     // X 승리인데 개수 안 맞음

    return 1; // 정상적인 상황
}

File: 프로그래머스/2/160585. 혼자서 하는 틱택토/혼자서 하는 틱택토.java
class Solution {
    public int solution(String[] board) {
        int countO = 0, countX = 0;

        // O, X 개수 세기
        for (String row : board) {
            for (char c : row.toCharArray()) {
                if (c == 'O') countO++;
                else if (c == 'X') countX++;
            }
        }

        boolean oWin = checkWin(board, 'O');
        boolean xWin = checkWin(board, 'X');

        // 1️⃣ X가 더 많으면 불가능
        if (countX > countO) return 0;
        // 2️⃣ O가 X보다 2개 이상 많으면 불가능
        if (countO > countX + 1) return 0;
        // 3️⃣ 둘 다 승리하면 불가능
        if (oWin && xWin) return 0;
        // 4️⃣ O가 이겼는데 개수가 안 맞으면 불가능
        if (oWin && countO != countX + 1) return 0;
        // 5️⃣ X가 이겼는데 개수가 안 맞으면 불가능
        if (xWin && countO != countX) return 0;

        // 모두 통과 시 가능한 상태
        return 1;
    }

    private boolean checkWin(String[] board, char player) {
        // 가로, 세로 검사
        for (int i = 0; i < 3; i++) {
            if (board[i].charAt(0) == player &&
                board[i].charAt(1) == player &&
                board[i].charAt(2) == player) return true;

            if (board[0].charAt(i) == player &&
                board[1].charAt(i) == player &&
                board[2].charAt(i) == player) return true;
        }

        // 대각선 검사
        if (board[0].charAt(0) == player &&
            board[1].charAt(1) == player &&
            board[2].charAt(2) == player) return true;

        if (board[0].charAt(2) == player &&
            board[1].charAt(1) == player &&
            board[2].charAt(0) == player) return true;

        return false;
    }
}
